### DISCUSSION TITLE
FOSS/Non MS intellisence extensions? or dropping VSCodium

### DISCUSSION THREAD

**Discussion by goyalyashpal** [VSCodium/vscodium#1278] [Q&A] [answered]
* NO offence to the VS codium team ofc, 
* it's more about MS enterprise crippling VSCodium, and not inherently about here. 
* i am just sharing how i feel - which is not very great
* and want your constructive opinion
* sorry for the semi-catchy title, it's not fake though... it's both possibilities as the `## action` section below also asks...

## Story
1. lots and lots of features depend on intellisense extensions (`ms-vscode.cpptools` and others)

    * autocomplete suggestions
    * language documentation?
    * find document usage/symbol definition (wording took from "geany editor")
    * and most latest being: sticky scroll
https://code.visualstudio.com/updates/v1_71#_sticky-scroll
https://code.visualstudio.com/updates/v1_70#_editor-sticky-scroll

2. i don't really want to use the extensions from MS, despite ~~them~~ their [upstream source](https://github.com/Microsoft/vscode-cpptools) being MIT licenced, due to the github copilot, and code search, and broken trust etc.

3. 
   * Without those aforementioned features, the main remaining irreplaceable features i use are just:
     + flawless multi-cursor/visual block mode
     + mind blowing workflow for viewing diffs - _**live**_ editable diff view, compare with saved, compare 2 selected open files, the explorer side tab, etc.
     + live `md` preview, 
   * with some things which are replaceable but just more polished in vscode being:
     + the whole find and replace functionality
     + reliable code highlighting for many formats i guess? 
   * With raaare-ities being 
     - ~~nice~~ theme support which seems to be widely used, 
     - command panel `^↑p`
     - the embedded terminal and i guess that's it.

   * Lots of things on the developer landscape are already under MS, and hence, intentionally don't play well with VSCodium (ghd, gh codespaces, and more to discover?) - so, if i have to use those, i am forced to use VSCode anyway.
   
   * It doesn't have flexible options like for its explorer tab, or even the "open with" button is typical "pulling back in to MS" type as it doesn't shows other editors or applications
   
   * - Being a heavy electron app, and that too from MS, and having all these unnecessary hurdles, 
      - I have grown lots of internal resistance and have next to none will remaining to use it. 
      - what more, it's been quite long since i last used it, and i am doing ohkay without it too. 
      - (ofcourse, i still exclusively use it for above mentioned 3 things - diff, flawless multi-cursor/block edit and occasionally for live md preview - as i've not found an alternate for these yet)

4. On the other hand, there are plenty other native editors (geany, kate, python IDLE, dev c++, etc.) have many of those "intellisense-dependent (in vscode)" features built in and some additional features which far outshine vscode.

## Action
+ So, considering all that, is there any alternates to those intellisense extensions which can rescue VSCodium for me?
+ or alternately: any other editor with good multi cursor support? 
don't confuse multi cursor and visual block editing though - they are separate thing. for example: even github instant editor (which launches on clicking "edit file" in a repo - has good multi cursor editing support.

**Comment by GitMensch** [comment, marked as answer]
> i don't really want to use the extensions from MS, despite their upstream source being MIT licenced

...haha... - https://github.com/microsoft/vscode-cpptools/blob/main/LICENSE.md says:
> Additional binary files may be downloaded at runtime after the initial installation; these are governed by the more restrictive proprietary license terms found at [RuntimeLicenses](RuntimeLicenses).

And all these "extended" features you reference are part of those binaries.

But the good news: there is no need to use vscode including VSCodium without language support.
I can recommend a bunch of ones, as you likely don't use COBOL may I suggest the clangd extension for c/c++ `llvm-vs-code-extensions.vscode-clangd` and ASM code lens `maziac.asm-code-lens` or `basdp.language-gas-x86` for assembler? Maybe additional `meronz.manpages` for easy access to the system documentation (also usable from Win32 specifying its path using MSYS2 or WSL).
For other languages there are also specific ones like `redhat.vscode-xml`, `redhat.java`, `alefragnani.pascal`, `XadillaX.viml`, `LaurentTreguier.rpm-spec`, ...

> considering all that, is there any alternates to those intellisense extensions which can rescue VSCodium for me?

Conclusion: yes.

**Reply by goyalyashpal** [reply]
thanks a lot for:
* giving the info on included binaries
* sharing the alternate extensions
* sharing the extensions id too

will be pasting the above paragraph list in itemised form in the following comment - modify ur original comment to be itemise if you so please (:

**Reply by goyalyashpal** [reply]
reminder: links extensions (ref: https://github.com/eclipse/openvsx/issues/363):
* `https://open-vsx.org/vscode/item?itemName=<ID>`
* `vscodium:extension/<ID>`

I suggest:

* `llvm-vs-code-extensions.vscode-clangd` : clangd for c/c++
* `maziac.asm-code-lens` : ASM code lens
* `basdp.language-gas-x86` : assembler?
* `meronz.manpages` : easy access to the system documentation (also usable from Win32 specifying its path using MSYS2 or WSL).

For other languages there are also specific ones like 

* `redhat.vscode-xml`
* `redhat.java`
* `alefragnani.pascal`
* `XadillaX.viml`
* `LaurentTreguier.rpm-spec`

**Reply by goyalyashpal** [reply]
aah, for using `clangd`, it will be a pain to put it in every folder: https://clangd.llvm.org/installation.html#project-setup

**Reply by GitMensch** [reply]
Depends, if you use any reasonable build system you can just use `bear make` once and have everything you need automatically. You likely want to build the source at least once in any case, no (when using cmake it is just one additional option passed to it)?

**Reply by goyalyashpal** [reply]
any recommendation for python?

from langserver.org, i could narrow down to these 2: [python-lsp-server] and other: [jedi-language-server]
`hx --health python` mentions `pylsp` which seems to correspond to the former one in above

(the red solid striken one being archived repos, and the blue dashed one being MS)
![image](https://user-images.githubusercontent.com/19423063/229374208-a7042991-5e17-450d-bde6-2f97eb0efeb2.png)


[python-lsp-server]: https://github.com/python-lsp/python-lsp-server/blob/714a5829192133f5fd7813a099741c580338c0bd/README.md?plain=1#L12-L15
[jedi-language-server]: https://github.com/pappasam/jedi-language-server

**Reply by goyalyashpal** [reply]
it seems things aren't good with python & vscodium, but maybe u know some way out...

* https://github.com/python-lsp/python-lsp-server/issues/39
* https://www.reddit.com/r/vscode/comments/uetxxt/pylsp_with_vscode/

**Reply by goyalyashpal** [reply]
continue this thread at: https://github.com/VSCodium/vscodium/discussions/1278#discussioncomment-5504011

**Comment by daiyam** [comment]
>> just more polished in vscode being:
> the whole find and replace functionality

I find that VSCode is lacking compared to jEdit...

The project could be better if more people were involved but FOSS proponent have a tendency to not like Microsoft project.

**Reply by goyalyashpal** [reply]
> The project could be better if more people were involved but FOSS proponent have a tendency to not like Microsoft project.

i like and want it better that way - kate on the alternate with similar modular organisation in backend - at least so does it looks from surface - seems better. the light theme and automatic yet customisable build menu of geany feels the best of all...

**Reply by goyalyashpal** [reply]
> I find that VSCode is lacking compared to jEdit...

thanks a lot for the recommendation - i'll be trying that too. how's it on the diff/multi-cursor editing front?

**Reply by daiyam** [reply]
@yashpalgoyal1304 I can't recommend jEdit anymore. I did use it for 15+ years before moving to VSCode. But if you really want, take a look at my settings (https://github.com/daiyam/jedit-settings), it will be a good start.
- `multi-cursor editing front`, I don't think it support that
- `diff`, not supported

**Reply by goyalyashpal** [reply]
if none of those two are not supported, and you are not rec-ing it, then there's no reason for me to use it either - i just was showing that i am eager to know about alternates (only gui ones for the time being)

i am thinking of looking out for contributing to kate someday in future - maybe u should try that too... got a feeling that community contribs can maybe more welcome there and hopefully will meet u in the dev chat there someday. cheers 👍

**Reply by goyalyashpal** [reply]
oh, and as for a quick summary on the landscape regarding many of the stuff i above mentioned:


multi-cursor editing:
* i have used this on github editor (web/online) first, and then on vscode (kinda web lol) - it works flawlessly at both places
* other editors **seem** to have only visual block mode, and NOT the arbitrary multi-cursor mode
* like: vim, notepad3 (confirmed)
* i am not sure about geany, kate - but i am leaning towards confirmation bias


diff

* i have seen the **live** i.e. on-the-go as you type diff viewer only in vscode and i guess it was gitlab **web**-IDE too
* non-live (IIRC) diff in a text editor : notepad++ (windows only) - surely there must be something for linux??
* diff in VCS : plenty in git clients


as for the diff-ing algorithm:
* i am yet to find a better diff algo that that of wikipedia (online) - yes, that rocks as far as at least non-code source is concerned
* it also highlights **moved** sections - wow


and same goes for the commit tree view 
* wikipedia's history view seems to outshine vscode stuff far behind
* i have heard that git clients do exist with advanced git tree features, but haven't researched about them personally yet.

**Reply by daiyam** [reply]
I use gitlens to improve the git experience on VSCodium.
For diff, I used https://sourcegear.com/diffmerge/ for quiet some time.

**Comment by goyalyashpal** [comment]
Waowwwww! [Found] a great tool, dedicated to visually diffing: [meld]

[found]: https://unix.stackexchange.com/questions/670458/670486#670486 "using-kate-text-editor-how-to-compare-two-files-and-find-differences"

- it also shows moved parts. - similar to the wikipedia's one. suuper awesome.
- it works in real time
- it can be triggered by kate (and so probably other tools too)

[meld]: https://meldmerge.org

**Reply by goyalyashpal** [reply]
https://user-images.githubusercontent.com/19423063/221839854-770320f4-5b6a-4c62-b6cf-f5c8e3596e40.mp4

**Reply by GitMensch** [reply]
Let's keep this discussion to vscodium, maybe it is even time to drop this discussion (= close it) ?

**Reply by goyalyashpal** [reply]
github discussions can't be closed yet... 😅

**Reply by GitMensch** [reply]
You're right, they would be "locked".

**Reply by goyalyashpal** [reply]
yes, they have added the close discussion button now.

![image](https://user-images.githubusercontent.com/19423063/226679802-37839e88-7313-41db-8291-36775e0d1754.png)

**Comment by goyalyashpal** [comment]
### Thread for python

continued from https://github.com/VSCodium/vscodium/discussions/1278#discussioncomment-5503912

well well, i disabled the extension `ms-python.python` and then the extension `almenon.arepl` (AREPL for python) started functioning correctly - whereas before it was showing similar error. But now, disabling ms-python, enabling arepl, then enabling ms-python; the arepl seems to be continuing to work as expected,...

> An Invalid Python interpreter is selected, please try changing it to enable features such as IntelliSense, linting, and debugging. See output for more details regarding why the interpreter is invalid.
\- `ms-python.python` extension 🙄 

The `formulahendry.code-runner` code runner is still not taking any input, but ah, it never really worked for me... will see

**Reply by goyalyashpal** [reply]
@GitMensch hi! any input for python LSP?

**Reply by GitMensch** [reply]
For my personal use `ms-python` on vscodium "pure" works fine as it falls back to the free-software LSP automatically and that is enough for my use, along with the python linters that are installed via pip if missing.

Note: on CentOS/RockyLinux/RHEL7 one needs to downgrade the extension version as it needs newer tools than normally available there (or install those from EPEL).

**Reply by goyalyashpal** [reply]
While searching for "static analyser python" in vscodium extensions:

* [pylyzer]\: `@id:pylyzer.pylyzer`
Python lang server writter in Rust, with type checking, invalid access detections like A-OOB & D-NEK[^invalid-access-errors]

[^invalid-access-errors]: A-OOB: Array Out of Bound, 
D-NEK: Dictionary Non-Existent Keys

* [Ruff]\: `@id:charliermarsh.ruff`
Python linter written in Rust

[pylyzer]: https://github.com/mtshiba/pylyzer "pylyzer static analyser: python lang-server"
[ruff]: https://github.com/charliermarsh/ruff "ruff static analyser: python linter"
[pylyzer.pylyzer]: 
[charliermarsh.ruff]:
